File: Data_Structures/Searching_and_Sorting/QuickSort.cpp
#include<stdio.h>
int partition(int a[],int beg,int end);
void quick_sort(int a[],int beg,int end);
int main()
{
    int arr[10],i,n,j,k;
    printf("Enter the number of elements in the array:");
    scanf("%d", &n);
    for(i=0;i<n;i++)
    {
        printf("arr[%d]= ",i);
        scanf("%d",&arr[i]);
    }
    quick_sort(arr,0,n-1);
    printf("The sorted array is:");
    for(i=0;i<n;i++)
        printf("%d\t",arr[i]);
}
int partition(int a[],int beg,int end)
{
    int left,right,temp,loc,flag=0;
    loc=left=beg;
    right=end;
    while(flag!=1)
    {
        while((a[loc]<=a[right])&&(loc!=right))
            right--;
        if(loc==right)
            flag=1;
        else if(a[loc]>a[right])
        {
            temp=a[loc];
            a[loc]=a[right];
            a[right]=temp;
            loc=right;
        }
        if(flag!=1)
        {
            while((a[loc]>=a[left])&&(loc!=left))
                left++;
            if(loc==left)
                flag=1;
            else if(a[loc]<a[left])
            {
                temp=a[loc];
                a[loc]=a[left];
                a[left]=temp;
                loc=left;
            }
        }
    }
    return loc;
}
void quick_sort(int a[],int beg,int end)
{
    int loc;
    if(beg<end)
    {
        loc=partition(a,beg,end);
        quick_sort(a,beg,loc-1);
        quick_sort(a,loc+1,end);
    }
}

File: Algorithms/Java/quickSort.java
// Java program for implementation of QuickSort 
import java.util.*;
class QuickSort 
{ 
	public static int partition(int arr[], int low, int high) 
	{ 
		int pivot = arr[high]; 
		int i = (low-1);  
		for (int j=low; j<high; j++) 
		{ 
			
			if (arr[j] < pivot) 
			{ 
				i++; 
				int temp = arr[i]; 
				arr[i] = arr[j]; 
				arr[j] = temp; 
			} 
		} 
		int temp = arr[i+1]; 
		arr[i+1] = arr[high]; 
		arr[high] = temp; 

		return i+1; 
	} 

	public static void sort(int arr[], int low, int high) 
	{ 
		if (low < high) 
		{ 
			
			int pi = partition(arr, low, high); 
			sort(arr, low, pi-1); 
			sort(arr, pi+1, high); 
		} 
	} 
	public static void printArray(int arr[]) 
	{ 
		int n = arr.length; 
		for (int i=0; i<n; ++i) 
			System.out.print(arr[i]+" "); 
		System.out.println(); 
	} 
	public static void main(String args[]) 
	{ 
		
                Scanner sc=new Scanner(System.in);
                System.out.println("Enter no. of array element :");
                int n=sc.nextInt();
		int ar[]=new int[n];
                System.out.println("Enter array elements: ");
                for(int i=0;i<n;i++)
                {
                    ar[i]=sc.nextInt();
                 }
                System.out.println("Array before sorting");
                printArray(ar); 
		sort(ar, 0, n-1); 
		System.out.println("sorted array"); 
		printArray(ar); 
	} 
} 

